libs/libc/unistd/lib_getoptvars.c

```diff
@@ -0,0 +1,170 @@
+
+/* Don't use TLS if we cannot avoid the memory leak or if TLS is disabled.
+ * Furthermore, TLS is not needed in the KERNEL build because it naturally
+ * supports per-process data.
+ */
+
+#if !defined(HAVE_DESTRUCTOR) && !defined(HAVE_CLEANUP)
+#  undef HAVE_TLSVARS
+#endif
+
+#ifdef CONFIG_BUILD_KERNEL
+#  undef HAVE_TLSVARS
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+#ifdef HAVE_TLSVARS
+static pthread_key_t g_getopt_key     = (pthread_key_t)INVALID_KEY;
+static pthread_mutex_t g_getopt_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+#else
+/* Fallback to avoid memory leak when there is no destructor and no cleanup
+ * or when there is no TLS.
+ */
+
+FAR struct getopt_s g_getopt_vars =
+{
+  NULL,
+  0,
+  1,
+  '?'
+};
+#endif
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+#ifdef HAVE_TLSVARS
+void _getopt_destructor(FAR void *getoptvars)
+{
+  lib_free(getoptvars);
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getoptvars
+ *
+ * Description:
+ *   Returns a pointer to to the thread-specific getopt() data.
+ *
+ ****************************************************************************/
+
+FAR struct getopt_s *getoptvars(void)
+{
+#ifdef HAVE_TLSVARS
+  FAR struct getopt_s *go;
+  int ret;
+
+  ret = pthread_mutex_lock(&g_getopt_mutex);
+  DEBUGASSERT(ret == OK);
+  UNUSED(ret);
+
+  /* Check if a key has been assigned */
+
+  if (!VALID_KEY(g_getopt_key))
+    {
+      ret = pthread_key_create(&g_getopt_key, _getopt_destructor);
+      DEBUGASSERT(ret == OK);
+      UNUSED(ret);
+    }
+
+  /* Get the structure of getopt() variables using the key. */
+
+  go = (FAR struct getopt_s *)pthread_getspecific(g_getopt_key);
+  if (go == NULL)
+    {
+      /* The instance has not yet been created.  Allocate getopt()
+       * it now.
+       */
+
+      go = (FAR struct getopt_s *)lib_malloc(sizeof(struct getopt_s));
+      DEBUGASSERT(go != NULL);
+
+      /* Initialize the getopt() variables */
+
+      go->go_optarg       = NULL;
+      go->go_opterr       = 0;
+      go->go_optind       = 1;
+      go->go_optopt       = '?';
+
+      go->go_optptr       = NULL;
+      go->go_binitialized = false;
+
+      /* Save this as the new thread-specific data */
+
+      ret = pthread_setspecific(g_getopt_key, go);
+
+#if !defined(HAVE_DESTRUCTOR) && defined(CONFIG_PTHREAD_CLEANUP)
+      if (ret == OK)
+        {
+          /* Free the thread-specific data when the thread exits */
+
+          pthread_cleanup_push(_getopt_destructor, go);
```

Comment:
from https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_cleanup_push.html:
pthread_cleanup_pop is desgined to be called before exit the function which call pthread_cleanup_push.
The usage in getopt will break the caller:
```
pthread_cleanup_push(my_destructor)
getopt
pthread_cleanup_pop-->execute _getopt_destructor not my_destructor
```

Reply:
I don't see the issue you are concerned with.

Reply:
Could you give more explanation here? pthread_cleanup_pop execute the wrong function and then postone my_destructor to the next up level(or even until thread exit). my_destructor may contain code to release mutex/semaphore, any delay may introduce the dead lock.

Reply:
> Could you give more explanation here? pthread_cleanup_pop execute the wrong function and then postone my_destructor to the next up level(or even until thread exit). my_destructor may contain code to release mutex/semaphore, any delay may introduce the dead lock.

Okay.  Now I understand.  Yes, I think that is a problem.  I don't see any solution other than the pthread_key_create() data destructor.  Implementing the basic destructor logic is not so difficult, but implementing a secure one the operates in user space is a lot of work.

Reply:
There is another problem with these pthread APIs.  Some will not work on the main thread.

pthread_cleanup_popall() is called from logic in pthread_exit().  However, I don't think there is anything in the exit() logic where pthread_cleanup_popall() is called.  The cleanup information is stored in the task TCB, but I don't think it is ever used.

The work on the features/pthread-user branch only addresses pthread_create(), pthread_startup(), and pthread_exit().  Additional logic would be needed to handle task exit().